README.md
## 微信跳一跳小游戏自动跳跃脚本


### 开发环境：

> Ubuntu 16

> IDEA 2017.3.4

> Node.js v9.5.0

> Npm v5.7.1

### 环境配置：

> [安装OpenCV](https://github.com/peterbraden/node-opencv)

> 下载本项目 ``git clone https://github.com/ChangedenCZD/wechat-auto-jumper.git``

> 初始化项目 ``npm install``

> 更新项目依赖 ``npm up -save``

> 运行项目 ``node ./src/index-2.0.js``

####
[ 更新日志 ](/CHANGE_LOG.md)
[ 意见反馈 ](https://github.com/ChangedenCZD/wechat-auto-jumper/issues)
[ 开源网站 ](https://www.chansos.com)
[ 其它开源项目 ](https://github.com/ChangedenCZD?tab=repositories)
####



CHANGE_LOG.md
#### v3.0
##### 2018-03-29
> 重写所有逻辑，大大提升了性能;

> 使用opencv4nodejs作为项目基础;

> 使用OpenCV的matchTemplate进行目标匹配;

> 待优化：准确率远远不如2.0（解决方案：添加更多box_template）

#### v2.0
##### 2018-03-08
> 优化坐标提取逻辑;

> 优化点击时长的算法;

#### v1.0
##### 2018-03-08
> 基于[wechat_jump_game](https://github.com/wangshub/wechat_jump_game/blob/master/wechat_jump_auto.py)项目的部分原理进行移植;

> 使用opencv处理原图，减少误差;


src/index-1.0.js
const cv = require('opencv');
const shell = require('shelljs');
const getPixels = require('get-pixels'); // 获取图片Buffer
const path = require('path');
const tesseract = require('node-tesseract'); // OCR
const lowThresh = 20;
const highThresh = 250;
const nIters = 0;

const RED = [0, 0, 255]; // B, G, R

const colorOffset = 100;

const SCREEN_CAPTURE_FILE_PATH = '/sdcard/wechat_jump.png';
const TARGET_FILE_PATH = '../images/src.png';
const CROP_FILE_PATH = '../images/crop.png';
const LEFT_PART_FILE_PATH = '../images/leftPart.png';
const RIGHT_PART_FILE_PATH = '../images/rightPart.png';

const DEBUG = true;

let timerId = null;

function run () {
    if (DEBUG) {
        jump();
    } else {
        let screenCaptureCommend = `adb shell screencap -p ${SCREEN_CAPTURE_FILE_PATH}`;
        let copyScreenCaptureCommend = `adb pull ${SCREEN_CAPTURE_FILE_PATH} ${TARGET_FILE_PATH}`;
        shell.exec(screenCaptureCommend);
        shell.exec(copyScreenCaptureCommend);
        tesseract.process(path.resolve(__dirname, TARGET_FILE_PATH), {
                l: 'chi_sim'
            }, function (err, text) {
                if (err) {
                    console.error(err);
                } else {
                    if (text.indexOf('最高分') > 0 || text.indexOf('排行榜') > 0) {//重开
                        console.log('重开');
                        swipe(554, 1584, 554, 1584, 317);
                        startTimer();
                    } else {
                        jump();
                    }
                }
            }
        );
    }
}

console.log('Is Debug?', DEBUG);
run();

function startTimer () {
    if (!DEBUG) {
        stopTimer();
        timerId = setTimeout(() => {
            run();
        }, 4000 + (genRandomNumber() % 10) * 1000);
    }
}

function stopTimer () {
    if (timerId) {
        clearTimeout(timerId);
    }
    timerId = null;
}

function jump () {
    cv.readImage(TARGET_FILE_PATH, function (err, im) {
        if (err) throw err;
        let width = im.width();
        let height = im.height();
        if (width < 1 || height < 1) throw new Error('Image has no size');

        let targetWidth = width;
        let targetHeight = height * 0.5;
        let im_canny = im.copy().crop(0, height * 0.25, targetWidth, targetHeight);
        im_canny.canny(lowThresh, highThresh);
        im_canny.dilate(nIters);
        let contours = im_canny.findContours();
        let outputFile = new cv.Matrix(targetHeight, targetWidth);
        outputFile.drawAllContours(contours, RED, -1);
        outputFile.save(CROP_FILE_PATH);
        let halfWidth = targetWidth * 0.5;
        let leftPart = outputFile.crop(0, 0, halfWidth, targetHeight);
        let rightPart = outputFile.crop(halfWidth, 0, halfWidth, targetHeight);
        leftPart.save(LEFT_PART_FILE_PATH);
        rightPart.save(RIGHT_PART_FILE_PATH);
        findTopPixel(LEFT_PART_FILE_PATH).then(result => {
            findPixelByX(LEFT_PART_FILE_PATH, result[0]).then(left => {
                findTopPixel(RIGHT_PART_FILE_PATH).then(result => {
                    findPixelByX(RIGHT_PART_FILE_PATH, result[0]).then(right => {
                        simulateSwipe(computeDistance(left, right, halfWidth), width, height);
                    });
                });
            });
        });
    });
}

function findTopPixel (file) {
    return new Promise(resolve => {
        getPixels(file, function (err, pixels) {
            if (err) {
                console.error(err);
            } else {
                let data = pixels.data;
                let targetWidth = pixels.shape[0];
                let targetHeight = pixels.shape[1];
                let length = targetWidth * targetHeight;
                let pixel = [0, 0];
                let t = 0;
                for (t; t < length; t++) {
                    let g = data[t * 4];
                    if (isValidColor(g)) {
                        let x = t % targetWidth;
                        if (isHorizontalLine(data, t, targetWidth)) {
                            t += targetWidth - x;
                        } else {
                            pixel = [x, parseInt(t / targetWidth)];
                            break;
                        }
                    }
                }
                console.log(pixel);
                resolve(pixel);
            }
        });
    });
}

function findPixelByX (file, x) {
    return new Promise(resolve => {
        getPixels(file, function (err, pixels) {
            if (err) {
                console.error(err);
            } else {
                let data = pixels.data;
                let targetWidth = pixels.shape[0];
                let targetHeight = pixels.shape[1];
                let length = targetWidth * targetHeight;
                let pixelList = [];
                let t = 0;
                for (t; t < length; t += targetWidth) {
                    let g = data[(t + x) * 4];
                    if (isValidColor(g) && !isHorizontalLine(data, t + x, targetWidth)) {
                        pixelList.push([(t + x) % targetWidth, parseInt(t / targetWidth)]);
                    }
                }
                resolve(pixelList);
            }
        });
    });
}

function isValidColor (color) {
    return color > colorOffset;
}

function isHorizontalLine (data, index) {
    let isLine = isValidColor(data[index * 4]);
    let validRange = index + 20;
    for (let i = index + 1; i < validRange; i++) {
        isLine = isLine && isValidColor(data[i * 4]);
        if (!isLine) {
            break;
        }
    }
    return isLine;
}

function computeDistance (left, right, offset) {
    let data = computePieceAndTarget(left, right, offset);
    let pieceIndex = data[0];
    let targetIndex = data[1];
    console.log(offset, pieceIndex, targetIndex, 'left ?', isPieceOnLeft(left, right));
    return Math.sqrt(Math.pow(Math.abs(pieceIndex[0] - targetIndex[0]), 2) + Math.pow(Math.abs(pieceIndex[1] - targetIndex[1]), 2));
}

function isPieceOnLeft (left, right) {
    let leftLikePiece = false;
    if (left.length >= 4) {
        let bodyHeaderRation = checkBodyHeaderRatio(left);
        leftLikePiece = bodyHeaderRation > 2 && bodyHeaderRation < 3;
    }
    let rightLikePiece = false;
    if (right.length >= 4) {
        let bodyHeaderRation = checkBodyHeaderRatio(right);
        rightLikePiece = bodyHeaderRation > 2 && bodyHeaderRation < 3;
    }
    return leftLikePiece || !rightLikePiece;
}

function checkBodyHeaderRatio (pixels) {
    try {
        return (pixels[3][1] - pixels[2][1]) / (pixels[1][1] - pixels[0][1]);
    } catch (e) {
        return 0;
    }
}

function computePieceAndTarget (left, right, offset) {
    let pieceIndex = [0, 0];
    let targetIndex = [0, 0];
    try {
        if (isPieceOnLeft(left, right)) { // 左边的是棋子
            pieceIndex = computePieceIndex(left);
            targetIndex = computeTargetIndex(right);
            targetIndex[0] += offset;
        } else {
            pieceIndex = computePieceIndex(right);
            targetIndex = computeTargetIndex(left);
            pieceIndex[0] += offset;
        }
    } catch (e) {//盲区
    }
    return [pieceIndex, targetIndex];
}

function computePieceIndex (pixels) {
    let topPixel = pixels[0];
    let topPixelY = topPixel[1];
    let height = pixels[3][1] - topPixelY;
    let bottomY = height * 190 / 210;
    let y = bottomY + topPixelY;
    return [topPixel[0], y];
}

function computeTargetIndex (pixels) {
    let topPixel = pixels[0];
    let bottomPixel = pixels[1];
    if (pixels.length > 4) {
        let a = bottomPixel[1] - topPixel[1];
        let b = pixels[3][1] - pixels[2][1];
        if (a <= b && (a + 40) >= b) {
            bottomPixel = pixels[3];
        }
    }
    return [topPixel[0], (bottomPixel[1] + topPixel[1]) / 2];
}

function simulateSwipe (distance, maxX, maxY) {
    let touchStartX = maxX / 2;
    let touchStartY = maxY / 2;
    touchStartX += genRandomNumber() % touchStartX;
    touchStartY += genRandomNumber() % touchStartY;
    let touchEndX = Math.min(genRandomNumber() % colorOffset + touchStartX, maxX);
    let touchEndY = Math.min(genRandomNumber() % colorOffset + touchStartY, maxY);
    let delay = Math.max(parseInt(distance * 1.5), 200);
    console.log(distance, delay);
    swipe(touchStartX, touchStartY, touchEndX, touchEndY, delay);
    startTimer();
}

function genRandomNumber () {
    return parseInt(Math.pow(Math.random() * 10, Math.random() * 10));
}

function swipe (touchStartX, touchStartY, touchEndX, touchEndY, delay) {
    let commend = `adb shell input swipe ${touchStartX} ${touchStartY} ${touchEndX} ${touchEndY} ${delay}`;
    console.log(commend);
    if (!DEBUG) {
        shell.exec(commend);
    }
}

src/index-3.0.js
const shell = require('shelljs');
const glob = require('glob');
const cv = require('opencv4nodejs');

const SCREEN_CAPTURE_FILE_PATH = '/sdcard/wechat_jump.jpg';
const TARGET_FILE_PATH = '../images/src.jpg';
const BOX_FILE_PATH = '../images/box.jpg';
const START_BUTTON_TEMPLATE_FILE_PATH = '../images/start_button_template.jpg';
const CLOSE_BUTTON_TEMPLATE_FILE_PATH = '../images/close_button_template.jpg';
const PIECE_TEMPLATE_FILE_PATH = '../images/piece_template.jpg';

const DEBUG = false;
const SCREEN_CAP = DEBUG ? false : true;

let startTime;
const pieceIndex = {x: 0, y: 0, position: 0};
const boxIndex = {x: 0, y: 0, position: 1};

let maxRow = 0;
let maxCol = 0;
let halfWidth = 0;
let halfHeight = 0;

let box = null;
let lastBox = null;
let distance = null;

let currentRound = 0;
let maxRound = 0;

function screenCap () {
    if (SCREEN_CAP) {
        let screenCaptureCommend = `adb shell screencap -p ${SCREEN_CAPTURE_FILE_PATH}`;
        let copyScreenCaptureCommend = `adb pull ${SCREEN_CAPTURE_FILE_PATH} ${TARGET_FILE_PATH}`;
        shell.exec(screenCaptureCommend);
        shell.exec(copyScreenCaptureCommend);
    }
}

function parseImage () {
    let image = cv.imread(TARGET_FILE_PATH);
    maxRow = image.rows;
    maxCol = image.cols;
    halfWidth = parseInt(maxCol * 0.5);
    halfHeight = parseInt(maxRow * 0.5);
    let mat = image.copy(new cv.Mat(maxRow, maxCol));
    box = mat.copy().gaussianBlur(new cv.Size(3, 3), 1).bgrToGray().canny(30, 200);
}

function findPiece () {
    let template = cv.imread(PIECE_TEMPLATE_FILE_PATH);
    template = template.bgrToGray();
    let match = box.matchTemplate(template, cv.TM_CCOEFF_NORMED);
    let loc = match.minMaxLoc();
    let maxLoc = loc.maxLoc;
    let x = pieceIndex.x = parseInt(maxLoc.x + template.cols * 0.5);
    let y = pieceIndex.y = parseInt(maxLoc.y + template.rows * 190 / 210);
    // let yOffset = 0.7;
    // let xOffset = 0.9;
    // let maxScanRow = parseInt(maxRow * yOffset);
    // let maxScanCol = parseInt(maxCol * xOffset);
    // let pieceScanStartX = parseInt(maxCol * (1 - xOffset));
    // let array = piece.getDataAsArray();
    // let headerIndex = null;
    // outer:
    //     for (let r = parseInt(maxRow * (1 - yOffset)); r < maxScanRow; r++) {
    //         let row = array[r];
    //         for (let c = pieceScanStartX; c < maxScanCol; c++) {
    //             let pixel = row[c];
    //             if (pixel < 255) {
    //                 if (!headerIndex) {
    //                     headerIndex = {x: c, y: r};
    //                     break outer;
    //                 }
    //             }
    //         }
    //     }
    // let x = pieceIndex.x = headerIndex.x + 5;
    // let y = pieceIndex.y = headerIndex.y + 190;
    if (x < 1 || y < 1) {
        console.log('Piece index error.');
        pieceIndex.position = -1;
    } else {
        pieceIndex.position = halfWidth > x ? 0 : 1;
    }
}

function findBox () {
    let piecePosition = pieceIndex.position;
    boxIndex.position = piecePosition > 0 ? 0 : 1;
    let targetIndex = null;
    glob.sync('../images/box_template/**/*.*').forEach(filePath => {
        targetIndex = compareIndex(targetIndex, findBoxByTemplate(filePath));
    });
    if (targetIndex && targetIndex.max < 0.5) {
        console.log('Compute Box Index For Pixel. TemplateIndex =', targetIndex);
        if (piecePosition >= 0) {
            let yOffset = 0.8;
            let xOffset = 0.9;
            let maxScanRow = parseInt(maxRow * yOffset);
            let boxScanStartX = 0;
            let maxScanCol = maxCol;
            let edge = 5;
            if (piecePosition === 0) { //棋子在左边
                boxScanStartX = parseInt(maxCol * (1 - xOffset));
                maxScanCol -= edge;
            } else {
                boxScanStartX += edge;
                maxScanCol = parseInt(maxCol * xOffset);
            }
            let array = box.getDataAsArray();
            let headerIndex = null;
            outer:
                for (let r = parseInt(maxRow * (1 - yOffset)); r < maxScanRow; r++) {
                    let row = array[r];
                    for (let c = boxScanStartX; c < maxScanCol; c++) {
                        let pixel = row[c];
                        if (pixel > 0) {
                            if (!headerIndex) {
                                headerIndex = {x: c, y: r};
                                break outer;
                            }
                        }
                    }
                }
            let nextY = tryToFindNextY(array, headerIndex);
            let x = headerIndex.x;
            let y = parseInt((headerIndex.y + nextY) * 0.5);
            targetIndex = {x, y};
        }
    }
    boxIndex.x = targetIndex.x;
    boxIndex.y = targetIndex.y + parseInt(maxRow * 0.25);
}

function findNextY (data, relativeIndex) {
    let nextY = relativeIndex.y;
    let x = relativeIndex.x;
    for (let r = nextY + 1; r < maxRow; r++) {
        if (data[r][x] > 0) {
            nextY = r;
            break;
        }
    }
    return nextY;
}

function tryToFindNextY (data, relativeIndex) {
    let y = relativeIndex.y;
    let x = relativeIndex.x;
    relativeIndex = {x, y};
    let nextYForRight;
    let nextYForLeft;
    let xOffset = 5;
    let maxX = Math.min(x + xOffset, maxRow);
    for (let c = x; c <= maxX; c++) {
        relativeIndex.x = c;
        nextYForRight = findNextY(data, relativeIndex);
        if (nextYForRight !== y) {
            break;
        }
    }
    let minX = Math.max(0, x - xOffset);
    for (let c = x; c >= minX; c--) {
        relativeIndex.x = c;
        nextYForLeft = findNextY(data, relativeIndex);
        if (nextYForLeft !== y) {
            break;
        }
    }
    let leftDifference = nextYForLeft - y;
    let value = nextYForLeft;
    if (leftDifference > 200 || leftDifference < 50) {
        value = nextYForRight;
    }
    return Math.max(y + 50, Math.min(value, y + 200));
}

function findBoxByTemplate (tempFilePath) {
    let template = cv.imread(tempFilePath);
    template = template.bgrToGray();
    let match = box.copy().matchTemplate(template, cv.TM_CCOEFF_NORMED);
    let loc = match.minMaxLoc();
    if (DEBUG) {
        console.log(tempFilePath, loc);
    }
    let maxLoc = loc.maxLoc;
    let x = parseInt(maxLoc.x + template.cols * 0.5);
    let y = parseInt(maxLoc.y + template.rows * 0.5);
    let max = loc.maxVal;
    return {x, y, max};
}

function compareIndex (lIndex, rIndex) {
    if (lIndex && rIndex) {
        return lIndex.max > rIndex.max ? lIndex : rIndex;
    } else if (lIndex) {
        return lIndex;
    }
    return rIndex;
}

function computeDistance () {
    distance = Math.sqrt(Math.pow(Math.abs(pieceIndex.x - boxIndex.x), 2) + Math.pow(Math.abs(pieceIndex.y - boxIndex.y), 2));
    // distance = Math.max(Math.min(distance, maxCol * 0.618), 200 / Math.sqrt(2));
}

function simulateSwipe () {
    if (distance) {
        currentRound++;
        let ratio = Math.sqrt(2);
        let delay = parseInt(Math.max(distance * ratio, 200));
        console.log('Piece index =', pieceIndex, 'Box index =', boxIndex, 'Distance =', distance, 'Delay =', delay);
        let touchStartX = parseInt(halfWidth * 0.5) + genRandomNumber() % halfWidth;
        let touchStartY = halfHeight + genRandomNumber() % halfHeight;
        let touchEndX = touchStartX + genRandomNumber() % 5;
        let touchEndY = touchStartY + genRandomNumber() % 5;
        swipe(touchStartX, touchStartY, touchEndX, touchEndY, delay);
        console.log(`尝试第 ${currentRound} 次跳跃`, `历史最高 ${maxRound} 次`);
        lastBox = box;
        tryToRestart(delay + 2000);
    }
}

function genRandomNumber () {
    return parseInt(Math.pow(Math.random() * 10, Math.random() * 10));
}

function swipe (touchStartX, touchStartY, touchEndX, touchEndY, delay) {
    let commend = `adb shell input swipe ${touchStartX} ${touchStartY} ${touchEndX || touchStartX} ${touchEndY || touchEndY} ${delay}`;
    console.log(commend);
    if (!DEBUG) {
        shell.exec(commend);
    }
}

function run () {
    startTime = Date.now();
    screenCap();
    parseImage();
    let template = cv.imread(CLOSE_BUTTON_TEMPLATE_FILE_PATH).bgrToGray();
    let match = box.matchTemplate(template, cv.TM_CCOEFF_NORMED);
    let closeButtonLoc = match.minMaxLoc();
    template = cv.imread(START_BUTTON_TEMPLATE_FILE_PATH).bgrToGray();
    match = box.matchTemplate(template, cv.TM_CCOEFF_NORMED);
    let startButtonLoc = match.minMaxLoc();
    if (startButtonLoc.maxVal > 0.5 && startButtonLoc.maxVal > closeButtonLoc.maxVal) {
        if (lastBox) {
            cv.imwrite(`../images/fail_box/${Date.now()}.jpg`, lastBox);
            console.log('记录失败模板');
        }
        lastBox = null;
        console.log('开始游戏');
        maxRound = Math.max(maxRound, currentRound);
        currentRound = 0;
        let maxLoc = startButtonLoc.maxLoc;
        let touchStartX = maxLoc.x + 50;
        let touchStartY = maxLoc.y + 10;
        let touchEndX = touchStartX + genRandomNumber() % 5;
        let touchEndY = touchStartY + genRandomNumber() % 5;
        swipe(touchStartX, touchStartY, touchEndX, touchEndY, 100);
        tryToRestart(800);
    } else if (closeButtonLoc.maxVal > 0.5) {
        console.log('关闭榜单');
        lastBox = null;
        let touchStartX = 979;
        let touchStartY = 297;
        let touchEndX = touchStartX + genRandomNumber() % 5;
        let touchEndY = touchStartY + genRandomNumber() % 5;
        swipe(touchStartX, touchStartY, touchEndX, touchEndY, 100);
        tryToRestart(800);
    } else {
        let rect = new cv.Rect(0, parseInt(maxRow * 0.25), maxCol, parseInt(maxRow * 0.5));
        box = box.getRegion(rect);
        if (DEBUG) {
            cv.imwrite(BOX_FILE_PATH, box);
        }
        findPiece();
        findBox();
        computeDistance();
        simulateSwipe();
    }
    console.log(`consume:${Date.now() - startTime}ms.`);
}

function tryToRestart (delay) {
    if (!DEBUG) {
        setTimeout(() => {
            run();
        }, delay);
    }
}

run();

src/index-2.0.js
const cv = require('opencv');
const shell = require('shelljs');
const getPixels = require('get-pixels'); // 获取图片Buffer
const path = require('path');
const tesseract = require('node-tesseract'); // OCR
const lowThresh = 20;
const highThresh = 195;
const nIters = 0;

const RED = [0, 0, 255]; // B, G, R

const colorOffset = 100;

const SCREEN_CAPTURE_FILE_PATH = '/sdcard/wechat_jump.png';
const TARGET_FILE_PATH = '../images/src.png';
const CROP_FILE_PATH = '../images/crop.png';
const LEFT_PART_FILE_PATH = '../images/leftPart.png';
const RIGHT_PART_FILE_PATH = '../images/rightPart.png';
const TEMP_FILE_PATH = '../images/temp.png';

const DEBUG = false;

let timerId = null;
let lastTargetLength = 0;
let count = 0;
let maxCount = 0;

function run () {
    if (DEBUG) {
        jump();
    } else {
        let screenCaptureCommend = `adb shell screencap -p ${SCREEN_CAPTURE_FILE_PATH}`;
        let copyScreenCaptureCommend = `adb pull ${SCREEN_CAPTURE_FILE_PATH} ${TARGET_FILE_PATH}`;
        shell.exec(screenCaptureCommend);
        shell.exec(copyScreenCaptureCommend);
        tesseract.process(path.resolve(__dirname, TARGET_FILE_PATH), {
                l: 'chi_sim'
            }, function (err, text) {
                if (err) {
                    console.error(err);
                } else {
                    if (text.indexOf('最高分') > 0 || text.indexOf('排行榜') > 0) {//重开
                        console.log('重开');
                        count = 0;
                        swipe(554, 1584, 554, 1584, 317);
                        startTimer();
                    } else {
                        jump();
                    }
                }
            }
        );
    }
}

console.log('Is Debug?', DEBUG);
run();

function startTimer () {
    if (!DEBUG) {
        stopTimer();
        let delay = 1444 + parseInt(((Math.random() * 100000) % 7000));
        timerId = setTimeout(() => {
            run();
        }, delay);
        console.log(`${delay} ms 后继续`);
    }
}

function stopTimer () {
    if (timerId) {
        clearTimeout(timerId);
    }
    timerId = null;
}

function jump () {
    cv.readImage(TARGET_FILE_PATH, function (err, im) {
        if (err) throw err;
        let width = im.width();
        let height = im.height();
        if (width < 1 || height < 1) throw new Error('Image has no size');

        let targetWidth = width;
        let targetHeight = height * 0.5;
        let im_canny = im.copy().crop(0, height * 0.25, targetWidth, targetHeight);
        im_canny.canny(lowThresh, highThresh);
        im_canny.dilate(nIters);
        let contours = im_canny.findContours();
        let outputFile = new cv.Matrix(targetHeight, targetWidth);
        outputFile.drawAllContours(contours, RED, -1);
        outputFile.save(CROP_FILE_PATH);
        let halfWidth = targetWidth * 0.5;
        let leftPart = outputFile.crop(0, 0, halfWidth, targetHeight);
        let rightPart = outputFile.crop(halfWidth, 0, halfWidth, targetHeight);
        leftPart.save(LEFT_PART_FILE_PATH);
        rightPart.save(RIGHT_PART_FILE_PATH);
        try {
            findIndex(LEFT_PART_FILE_PATH).then(left => {
                findPixelByX(LEFT_PART_FILE_PATH, left[0]).then(left => {
                    findIndex(RIGHT_PART_FILE_PATH).then(right => {
                        findPixelByX(RIGHT_PART_FILE_PATH, right[0]).then(right => {
                            count++;
                            maxCount = Math.max(count, maxCount);
                            console.log(`第${count}次跳跃，最高次数${maxCount}次`);
                            simulateSwipe(computeDistance(left, right, halfWidth), width, height);
                        });
                    });
                });
            });
        } catch (e) {
            console.error(e);
            startTimer();
        }
    });
}

function findIndex (filePath) {
    return new Promise(resolve => {
        checkAndFixImage(filePath, 0).then(result => {
            resolve(result);
        });
    });
}

function checkAndFixImage (filePath, fixCount, cb) {
    return new Promise(resolve => {
        cb = cb || resolve;
        findTopPixel(filePath).then(result => {
            let pixel = result.pixel;
            if (isValidTop(pixel, result.image)) {
                pixel[0] += fixCount;
                cb(pixel);
            } else {
                fixImage(filePath).then(() => {
                    fixCount++;
                    checkAndFixImage(TEMP_FILE_PATH, fixCount, cb);
                });
            }
        });
    });
}

function fixImage (filePath) {
    return new Promise(resolve => {
        cv.readImage(filePath, function (err, im) {
            if (err) throw err;
            let width = im.width();
            let height = im.height();
            if (width < 1 || height < 1) throw new Error('Image has no size');
            let im_canny = im.copy().crop(1, 0, width - 2, height);
            im_canny.save(TEMP_FILE_PATH);
            resolve();
        });
    });
}

function isValidTop (pixel, image) {
    let data = image.data;
    let targetWidth = image.shape[0];
    let x = pixel[0];
    return x > 0 && !((pixel[0] + 5) > targetWidth && isValidColor(data[((pixel[1] + 1) * targetWidth - 1) * 4]));
}

function findTopPixel (file) {
    return new Promise(resolve => {
        getPixels(file, function (err, image) {
            if (err) {
                console.error(err);
            } else {
                let data = image.data;
                let targetWidth = image.shape[0];
                let targetHeight = image.shape[1];
                let length = targetWidth * targetHeight;
                let pixel = [0, 0];
                let t = 0;
                for (t; t < length; t++) {
                    let g = data[t * 4];
                    if (isValidColor(g)) {
                        let x = t % targetWidth;
                        pixel = [x, parseInt(t / targetWidth)];
                        break;
                    }
                }
                resolve({
                    pixel, image
                });
            }
        });
    });
}

function findPixelByX (file, x) {
    return new Promise(resolve => {
        getPixels(file, function (err, pixels) {
            if (err) {
                console.error(err);
            } else {
                let data = pixels.data;
                let targetWidth = pixels.shape[0];
                let targetHeight = pixels.shape[1];
                let length = targetWidth * targetHeight;
                let pixelList = [];
                let t = 0;
                for (t; t < length; t += targetWidth) {
                    let g = data[(t + x) * 4];
                    if (isValidColor(g)) {
                        pixelList.push([(t + x) % targetWidth, parseInt(t / targetWidth)]);
                    }
                }
                resolve(pixelList);
            }
        });
    });
}

function isValidColor (color) {
    return color > colorOffset;
}

function computeDistance (left, right, offset) {
    let data = computePieceAndTarget(left, right, offset);
    let pieceIndex = data[0];
    let targetIndex = data[1];
    console.log('PieceIndex =', pieceIndex, 'TargetIndex=', targetIndex, 'Piece on left of center?', isPieceOnLeft(left, right));
    return Math.sqrt(Math.pow(Math.abs(pieceIndex[0] - targetIndex[0]), 2) + Math.pow(Math.abs(pieceIndex[1] - targetIndex[1]), 2));
}

function isPieceOnLeft (left, right) {
    let leftLikePiece = false;
    if (left.length >= 4) {
        let bodyHeaderRation = checkBodyHeaderRatio(left);
        leftLikePiece = bodyHeaderRation > 2 && bodyHeaderRation < 3;
    }
    let rightLikePiece = false;
    if (right.length >= 4) {
        let bodyHeaderRation = checkBodyHeaderRatio(right);
        rightLikePiece = bodyHeaderRation > 2 && bodyHeaderRation < 3;
    }
    return leftLikePiece || !rightLikePiece;
}

function checkBodyHeaderRatio (pixels) {
    try {
        return (pixels[3][1] - pixels[2][1]) / (pixels[1][1] - pixels[0][1]);
    } catch (e) {
        return 0;
    }
}

function computePieceAndTarget (left, right, offset) {
    let pieceIndex = [0, 0];
    let targetIndex = [0, 0];
    try {
        if (isPieceOnLeft(left, right)) { // 左边的是棋子
            pieceIndex = computePieceIndex(left);
            targetIndex = computeTargetIndex(right);
            targetIndex[0] += offset;
        } else {
            pieceIndex = computePieceIndex(right);
            targetIndex = computeTargetIndex(left);
            pieceIndex[0] += offset;
        }
    } catch (e) {//盲区
        console.error(e);
        throw e;
    }
    pieceIndex[0] += 3;
    return [pieceIndex, targetIndex];
}

function computePieceIndex (pixels) {
    let topPixel = pixels[0];
    let topPixelY = topPixel[1];
    let height = pixels[3][1] - topPixelY;
    let bottomY = height * 190 / 210;
    let y = bottomY + topPixelY;
    return [topPixel[0], y];
}

function computeTargetIndex (pixels) {
    let topPixel = pixels[0];
    let bottomPixel = pixels[1];
    let a = bottomPixel[1] - topPixel[1];
    if (a < 11) {
        topPixel = pixels[1];
        bottomPixel = pixels[2];
    } else if (a < 200 && pixels.length > 4) {
        let a = bottomPixel[1] - topPixel[1];
        let b = pixels[3][1] - pixels[2][1];
        if (b >= a && (pixels[2][1] - pixels[1][1]) < a && (a + 40) >= b) {
            bottomPixel = pixels[3];
        }
    }
    console.log('Target first distance =', a);
    lastTargetLength = pixels.length;
    return [topPixel[0], (bottomPixel[1] + topPixel[1]) / 2];
}

function simulateSwipe (distance, maxX, maxY) {
    let touchStartX = maxX / 2;
    let touchStartY = maxY / 2;
    touchStartX += genRandomNumber() % touchStartX;
    touchStartY += genRandomNumber() % touchStartY;
    let touchEndX = Math.min(genRandomNumber() % colorOffset + touchStartX, maxX);
    let touchEndY = Math.min(genRandomNumber() % colorOffset + touchStartY, maxY);
    let ratio = Math.sqrt(2 + (lastTargetLength > 4 ? -0.1 : 0));
    let delay = Math.max(parseInt(distance * ratio), 200);
    console.log('distance =', distance, 'delay =', delay, 'ratio =', ratio);
    swipe(touchStartX, touchStartY, touchEndX, touchEndY, delay);
    startTimer();
}

function genRandomNumber () {
    return parseInt(Math.pow(Math.random() * 10, Math.random() * 10));
}

function swipe (touchStartX, touchStartY, touchEndX, touchEndY, delay) {
    let commend = `adb shell input swipe ${touchStartX} ${touchStartY} ${touchEndX} ${touchEndY} ${delay}`;
    console.log(commend);
    if (!DEBUG) {
        shell.exec(commend);
    }
}